### ISSUE TITLE
Huge performance drop when a directional light is rotating

### ISSUE THREAD

**Issue by moisesjpelaez** [godotengine/godot#53969]
### Godot version

Godot Engine v4.0.dev.20211015.official.f113dc986

### System information

Windows 10, Vulkan API 1.2.162, AMD Radeon R7 Graphics

### Issue description

When I open a small project I created and test it everything runs very smooth, even smoother than in Godot 3.x versions, however when I rotate a directional light I have a huge lag. I tried both in the editor and making a simple script to rotate the light in game.

Light without rotating video: https://user-images.githubusercontent.com/24682046/137772817-0da6143e-9565-4cd0-bd4e-2bc3dba69f73.mp4
Light rotating video: https://user-images.githubusercontent.com/24682046/137772705-9ce3468f-b404-44bb-9bce-1b3b7011ef5c.mp4

(you can see the FPS count on top left)

### Steps to reproduce

1) Setup a basic scene with a cube, a plane and default environment
2) Add a directional light
3) Rotate it in the editor

or

3) Add a simple script to rotate the light inside the `_process` function
4) Play the scene

### Minimal reproduction project

[Godot4Test.zip](https://github.com/godotengine/godot/files/7367079/Godot4Test.zip)

**Comment by moisesjpelaez** [comment]
> I can reproduce this on commit [c240275](https://github.com/godotengine/godot/commit/c2402754821e8fa6a6666531c9dbbcd2193d4e05) (Linux, GeForce GTX 1080 with NVIDIA 470.74 driver).
> 
> With V-Sync disabled, I get 1,300+ FPS (0.8 mspf) when the light is not rotated, and only 167 FPS (6.0 mspf) when the light is rotated every frame. However, I get 604 FPS (1.7 mspf) when the Environment's sky update mode is set to **RealTime** instead of the default **Automatic**. This update mode will make sharp reflections look worse in general, but it's better-suited to day/night cycles.
> 
> I suggest trying this update mode locally. It won't be as fast as not rotating the light at all (since the light is included in the sky and automatically rotated now), but it should be fast enough.
> 
> ![image](https://user-images.githubusercontent.com/180032/137777318-350b485c-cc2b-4ca2-9019-2229adb25bc8.png)
> 
> PS: For other people testing the MRP, make sure to attach the `RotateObject.gd` script to the DirectionalLight3D node.

I tried this and got better performance but it starts at 49~52 FPS then drops to 39~44 FPS after some seconds.

**Comment by clayjohn** [comment]
Indeed, the solution here is to use the ``RealTime`` process mode. The default process mode does the expensive version of the cubemap update over several frames, but still is quite resource intensive and is not intended to be used for frequent real time updates. 

> I tried this and got better performance but it starts at 49 ~ 52 FPS then drops to 39 ~ 44 FPS after some seconds.

Could you clarify what this means? Is this a drop during a constant rotation? Or is it dropping during a time the light isn't rotating?

**Comment by moisesjpelaez** [comment]
The drop is during a constant rotation, with the `RotateObject.gd` script attached to the light.

**Comment by tavurth** [comment]
Just ran across this yesterday.

As far as I've read that means the code which can be found here is at fault:

https://github.com/godotengine/godot/blob/2362c5806773aca4a728a3f50a9de0f568bdac73/servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp#L333-L334

This code seems to be called here:

https://github.com/godotengine/godot/blob/11cefc2b28faad9ac05f0ffccbe3b77f618bc2a1/servers/rendering/renderer_rd/renderer_scene_render_rd.cpp#L759-L764

So the fix here would be either to buffer calling that update function, or to buffer calling the clear function?

https://github.com/godotengine/godot/blob/2362c5806773aca4a728a3f50a9de0f568bdac73/servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp#L656-L673

**Comment by tavurth** [comment]
This is currently rebuilding everything at 144hz, perhaps that is something that users could want if they run a timelapse sunrise, but I would guess than in most cases the game clock will advance much more slowly (rotating the light much more slowly)

We should also consider how fast the sun is rotating, I feel like 144hz of `0.01 rads` is still going to create an effect where the sun is moving far too fast for any meaningful timelapse.

In this case we can reject the idea that most (if anyone) needs to update their lightmap that fast. I will create a short video to demonstrate.

**Comment by tavurth** [comment]
Unfortunately on my device, this issue causes me to only reach 20fps and so I cannot check for smooth interpolation of the camera.

https://user-images.githubusercontent.com/100964/153867158-68532d73-2b08-46d5-819e-ec732989a12f.mov

Perhaps someone else can try this:

[godot-directionallight3d.zip](https://github.com/godotengine/godot/files/8060808/godot-directionallight3d.zip)

I would say that preferring performance for low end devices would make more sense in this case, as whenever the light changes in a game with a dynamic sky, all mobile & low end devices will suffer jankiness & frame skipping anyway.

**Comment by Zireael07** [comment]
No, I don't think we need to do it at 144hz, do we? :confused:
